src/lib/marlin_plonk_bindings/wasm/src/pasta_fq.rs

```diff
@@ -0,0 +1,200 @@
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::convert::{OptionIntoWasmAbi, IntoWasmAbi, FromWasmAbi};
+use crate::bigint_256::{self, WasmBigInteger256};
+use algebra::biginteger::BigInteger256;
+use algebra::{ToBytes, FromBytes};
+use mina_curves::pasta::fq::{Fq, FqParameters as Fq_params};
+use algebra::{
+    fields::{Field, FpParameters, PrimeField, SquareRootField},
+    FftField, One, UniformRand, Zero,
+};
+use ff_fft::{EvaluationDomain, Radix2EvaluationDomain as Domain};
+use num_bigint::BigUint;
+use rand::rngs::StdRng;
+use std::cmp::Ordering::{Equal, Greater, Less};
+
+pub struct WasmPastaFq(pub Fq);
+
+impl wasm_bindgen::describe::WasmDescribe for WasmPastaFq {
+    fn describe() { <Vec<u8> as wasm_bindgen::describe::WasmDescribe>::describe() }
+}
+
+impl FromWasmAbi for WasmPastaFq {
+    type Abi = <Vec<u8> as FromWasmAbi>::Abi;
+    unsafe fn from_abi(js: Self::Abi) -> Self {
+        let bytes: Vec<u8> = FromWasmAbi::from_abi(js);
+        WasmPastaFq(FromBytes::read(bytes.as_slice()).unwrap())
+    }
+}
+
+impl IntoWasmAbi for WasmPastaFq {
+    type Abi = <Vec<u8> as FromWasmAbi>::Abi;
+    fn into_abi(self) -> Self::Abi {
+        let mut bytes: Vec<u8> = vec![];
+        self.0.write(&mut bytes);
+        bytes.into_abi()
+    }
+}
+
+impl OptionIntoWasmAbi for WasmPastaFq {
+    fn none() -> Self::Abi {
+        let max_bigint = WasmBigInteger256(BigInteger256([u64::MAX, u64::MAX, u64::MAX, u64::MAX]));
+        max_bigint.into_abi()
+    }
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_size_in_bits() -> isize {
+    Fq_params::MODULUS_BITS as isize
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_size() -> WasmBigInteger256 {
+    WasmBigInteger256(Fq_params::MODULUS)
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_add(x: WasmPastaFq, y: WasmPastaFq) -> WasmPastaFq {
+    WasmPastaFq(x.0 + y.0)
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_sub(x: WasmPastaFq, y: WasmPastaFq) -> WasmPastaFq {
+    WasmPastaFq(x.0 - y.0)
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_negate(x: WasmPastaFq) -> WasmPastaFq {
+    WasmPastaFq(-x.0)
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_mul(x: WasmPastaFq, y: WasmPastaFq) -> WasmPastaFq {
+    WasmPastaFq(x.0 * y.0)
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_div(x: WasmPastaFq, y: WasmPastaFq) -> WasmPastaFq {
+    WasmPastaFq(x.0 / y.0)
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_inv(x: WasmPastaFq) -> Option<WasmPastaFq> {
+    x.0.inverse().map(|x| { WasmPastaFq(x) })
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_square(x: WasmPastaFq) -> WasmPastaFq {
+    WasmPastaFq(x.0.square())
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_is_square(x: WasmPastaFq) -> bool {
+    let s = x.0.pow(Fq_params::MODULUS_MINUS_ONE_DIV_TWO);
+    s.is_zero() || s.is_one()
+}
+
+#[wasm_bindgen]
+pub fn caml_pasta_fq_sqrt(x: WasmPastaFq) -> Option<WasmPastaFq> {
+    x.0.sqrt().map(|x| { WasmPastaFq(x) })
+}
+
+#[wasm_bindgen]
```

Comment:
We don't have to do it now, but for this sort of thing in the future do you think it's worth using macros (since we have the same thing copy pasted but with q instead of p)?

Reply:
Yea, this would be good to do in the future

Reply:
I was wondering the same in the marlin_plonk_bindings, lots of code we could dedup